uaclient/cli.py

```diff
@@ -197,155 +272,289 @@ def _perform_enable(entitlement_name: str, cfg: config.UAConfig, *,
     return ret
 
 
-@assert_attached_root
+@assert_root
+@require_valid_entitlement_name("enable")
+@assert_attached(ua_status.MESSAGE_ENABLE_FAILURE_UNATTACHED_TMPL)
 def action_enable(args, cfg):
     """Perform the enable action on a named entitlement.
 
     @return: 0 on success, 1 otherwise
     """
     print(ua_status.MESSAGE_REFRESH_ENABLE)
-    if not contract.request_updated_contract(cfg):
-        logging.debug(ua_status.MESSAGE_REFRESH_FAILURE)
+    try:
+        contract.request_updated_contract(cfg)
+    except (util.UrlError, exceptions.UserFacingError):
+        # Inability to refresh is not a critical issue during enable
+        logging.debug(ua_status.MESSAGE_REFRESH_FAILURE, exc_info=True)
     return 0 if _perform_enable(args.name, cfg) else 1
 
 
-@assert_attached_root
+@assert_root
+@assert_attached()
 def action_detach(args, cfg):
     """Perform the detach action for this machine.
 
     @return: 0 on success, 1 otherwise
     """
+    to_disable = []
     for ent_cls in entitlements.ENTITLEMENT_CLASSES:
         ent = ent_cls(cfg)
         if ent.can_disable(silent=True):
-            ent.disable(silent=True)
+            to_disable.append(ent)
+    if to_disable:
+        suffix = "s" if len(to_disable) > 1 else ""
+        print("Detach will disable the following service{}:".format(suffix))
+        for ent in to_disable:
+            print("    {}".format(ent.name))
+    if not util.prompt_for_confirmation():
+        return 1
+    for ent in to_disable:
+        ent.disable(silent=True)
     cfg.delete_cache()
     print(ua_status.MESSAGE_DETACH_SUCCESS)
     return 0
 
 
-def action_attach(args, cfg):
+def action_attach_premium(args, cfg):
+    from uaclient.clouds import identity
+
     if cfg.is_attached:
-        print("This machine is already attached to '%s'." %
-              cfg.accounts[0]['name'])
+        print(
+            "This machine is already attached to '{}'.".format(
+                cfg.accounts[0]["name"]
+            )
+        )
         return 0
     if os.getuid() != 0:
         raise exceptions.NonRootUserError()
-    contract_token = args.token
-    if not contract_token:
-        print('No valid contract token available')
-        return 1
+    cloud_type = identity.get_cloud_type()
+    if cloud_type != "aws":
+        print(
+            "No premium image support for cloud '%s'. Use ua attach <token>"
```

Comment:
Change to:
"Premium image support is not supported on '%s'
For more information see: https://ubuntu.com/advantage"

Ensure it is on two lines.